File: _assets/scripts/normalizers/models_model_normalize.lua
return function (data)
    local data_attr = python.as_attrgetter(data)
    local output = python.builtins.dict()
    local output_attr = python.as_attrgetter(output)
    for index1, key_value1 in python.enumerate(data_attr.items()) do
        local model_file_name = key_value1[0]
        local resource_packs = key_value1[1]
        local resource_packs_attr = python.as_attrgetter(resource_packs)
        for index2, key_value2 in python.enumerate(resource_packs_attr.items()) do
            local resource_pack_name = key_value2[0]
            local model_file_data = key_value2[1]
            local model_file_data_attr = python.as_attrgetter(model_file_data)
            if model_file_data_attr.__contains__("minecraft:geometry") then
                local format_version = model_file_data["format_version"]
                assert(python.builtins.type(model_file_data["minecraft:geometry"] == python.builtins.list))
                for index3, geometry_item in python.enumerate(model_file_data["minecraft:geometry"]) do
                    local name = geometry_item["description"]["identifier"]
                    local model_output_name = model_file_name .. " " .. name
                    if not output_attr.__contains__(model_output_name) then
                        output[model_output_name] = python.builtins.dict()
                    end
                    output[model_output_name][resource_pack_name] = python.builtins.dict()
                    output[model_output_name][resource_pack_name]["format_version"] = format_version
                    output[model_output_name][resource_pack_name]["minecraft:geometry"] = geometry_item
                end
            else
                local format_version = model_file_data_attr.get("format_version")
                for index3, key_value3 in python.enumerate(model_file_data_attr.items()) do
                    local name = key_value3[0]
                    local model_data = key_value3[1]
                    if name ~= "format_version" then
                        local model_data_attr = python.as_attrgetter(model_data)
                        local description_dict = python.builtins.dict()
                        description_dict["identifier"] = name
                        local model_output_name = model_file_name .. " " .. name
                        local keys = {}
                        for index4, description_key in python.enumerate(model_data_attr.keys()) do
                            if description_key ~= "bones" then
                                table.insert(keys, description_key)
                            end
                        end
                        -- split into two for statements because dicts can't have keys deleted while iterating.
                        for index4, description_key in pairs(keys) do
                            description_dict[description_key] = model_data[description_key]
                            model_data_attr.__delitem__(description_key)
                        end
                        model_data["description"] = description_dict
                        if not output_attr.__contains__(model_output_name) then
                            output[model_output_name] = python.builtins.dict()
                        end
                        output[model_output_name][resource_pack_name] = python.builtins.dict()
                        output[model_output_name][resource_pack_name]["format_version"] = format_version
                        output[model_output_name][resource_pack_name]["minecraft:geometry"] = model_data
                    end
                end
            end
        end
    end
    return output
end

File: _assets/scripts/normalizers/models_model_normalize.py
from collections import defaultdict
from typing import Any, NotRequired, TypedDict, cast

import Utilities.File as File

__all__ = ["models_model_normalize"]

description_typed_dict = TypedDict("description_typed_dict", {
    "identifier": str,
    "texture_height": NotRequired[Any],
    "texture_width": NotRequired[Any],
    "textureheight": NotRequired[Any],
    "texturewidth": NotRequired[Any],
    "visible_bounds_height": NotRequired[Any],
    "visible_bounds_offset": NotRequired[Any],
    "visible_bounds_width": NotRequired[Any],
})

geometry_typed_dict = TypedDict("geometry_typed_dict", {
    "description": description_typed_dict,
    "bones": dict[str,Any]
})

model_file_type1 = TypedDict("model_file_type1", {"minecraft:geometry": list[geometry_typed_dict], "format_version": str})

model2_model_typed_dict = TypedDict("model2_model_typed_dict", {
    "description": NotRequired["model2_model_typed_dict"], # so it doesn't yell at me when I assign this key
    "identifier": str,
    "bones": dict[str,Any],
    "texture_height": NotRequired[Any],
    "texture_width": NotRequired[Any],
    "textureheight": NotRequired[Any],
    "texturewidth": NotRequired[Any],
    "visible_bounds_height": NotRequired[Any],
    "visible_bounds_offset": NotRequired[Any],
    "visible_bounds_width": NotRequired[Any],
})

model_file_type2 = dict[str,model2_model_typed_dict]

output_typed_dict = TypedDict("output_typed_dict", {"minecraft:geometry": geometry_typed_dict, "format_version": str})

def models_model_normalize(data:dict[str,dict[str,File.File[model_file_type1|model_file_type2]]]) -> File.FakeFile[dict[str,dict[str,output_typed_dict]]]:
    output:dict[str,dict[str,output_typed_dict]] = defaultdict(lambda: {})
    file_hashes:list[int] = []
    for model_file_name, resource_packs in data.items():
        for resource_pack_name, model_file in resource_packs.items():
            file_hashes.append(hash(model_file))
            model_file_data = model_file.data
            if "minecraft:geometry" in model_file_data:
                model_file_data = cast(model_file_type1, model_file_data)
                format_version = model_file_data["format_version"]
                for geometry_item in model_file_data["minecraft:geometry"]:
                    name = geometry_item["description"]["identifier"]
                    model_output_name = "%s %s" % (model_file_name, name)
                    output[model_output_name][resource_pack_name] = {
                        "format_version": format_version,
                        "minecraft:geometry": geometry_item,
                    }
            else:
                format_version = cast(str, model_file_data.get("format_version"))
                for name, model_data in model_file_data.items():
                    if name == "format_version": continue
                    description:description_typed_dict = {
                        "identifier": name,
                    }
                    model_output_name = "%s %s" % (model_file_name, name)
                    for description_key in [key for key in model_data.keys() if key != "bones"]:
                        description[description_key] = model_data[description_key]
                        del model_data[description_key]
                    model_data = cast(geometry_typed_dict, model_data)
                    model_data["description"] = description
                    output[model_output_name][resource_pack_name] = {
                        "format_version": format_version,
                        "minecraft:geometry": model_data,
                    }
    return File.FakeFile("combined_models_file", output, hash(tuple(file_hashes)))
